src/tests/test_run.py

```diff
@@ -99,85 +87,82 @@ def test_handle_postgres_special(mock_conns):
-
-    with warnings.catch_warnings():
-        warnings.simplefilter("error")
-
-    assert "The database driver doesn't support such " in caplog.records[0].msg
-    assert output is True
-
-
-def test_pytds_autocommit(pytds_conns, mock_config):
-    with warnings.catch_warnings(record=True) as w:
-        output = set_autocommit(pytds_conns, mock_config)
-        assert (
-            str(w[-1].message)
-            == "Autocommit is not supported for pytds, thus is automatically disabled"
-        )
-        assert output is False
-
-
-def test_select_df_type_is_pandas(monkeypatch, config_pandas, mock_resultset):
-    monkeypatch.setattr("sql.run.select_df_type", mock_resultset.DataFrame())
-    output = select_df_type(mock_resultset, config_pandas)
+def test_select_df_type_is_pandas(mock_resultset):
+    output = select_df_type(mock_resultset, ConfigPandas)
     assert isinstance(output, pandas.DataFrame)
 
 
-def test_select_df_type_is_polars(monkeypatch, config_polars, mock_resultset):
-    monkeypatch.setattr("sql.run.select_df_type", mock_resultset.PolarsDataFrame())
-    output = select_df_type(mock_resultset, config_polars)
+def test_select_df_type_is_polars(mock_resultset):
+    output = select_df_type(mock_resultset, ConfigPolars)
     assert isinstance(output, polars.DataFrame)
 
 
-def test_sql_starts_with_begin(mock_conns, mock_config):
+def test_sql_starts_with_begin(mock_conns):
     with pytest.raises(UsageError, match="does not support transactions") as excinfo:
-        run(mock_conns, "BEGIN", mock_config)
+        run_statements(mock_conns, "BEGIN", Config)
 
     assert excinfo.value.error_type == "RuntimeError"
 
 
-def test_sql_is_empty(mock_conns, mock_config):
-    assert run(mock_conns, "  ", mock_config) == "Connected: %s" % mock_conns.name
+def test_sql_is_empty(mock_conns):
+    assert run_statements(mock_conns, "  ", Config) == "Connected: %s" % mock_conns.name
 
 
-def test_run(monkeypatch, mock_conns, mock_resultset, config_pandas):
-    monkeypatch.setattr("sql.run.handle_postgres_special", Mock())
-    monkeypatch.setattr("sql.run._commit", Mock())
-    monkeypatch.setattr("sql.run.display_affected_rowcount", Mock())
-    monkeypatch.setattr("sql.run.ResultSet", mock_resultset)
-
-    output = run(mock_conns, "\\", config_pandas)
-    assert isinstance(output, type(mock_resultset.DataFrame()))
+# @pytest.mark.parametrize(
+#     "n, message",
+#     [
+#         [1, "1 rows affected.\n"],
+#         [0, ""],
+#     ],
+# )
+# def test_display_affected_rowcount(capsys, n, message):
+#     display_affected_rowcount(n)
+#     captured = capsys.readouterr()
+#     assert captured.out == message
 
 
 @pytest.mark.parametrize(
-    "n, message",
+    "connection",
     [
-        [1, "1 rows affected.\n"],
-        [0, ""],
+        SQLAlchemyConnection(create_engine("duckdb://")),
+        SQLAlchemyConnection(create_engine("sqlite://")),
+        DBAPIConnection(duckdb.connect()),
```

Comment:
A bit confused here. Why are we testing DBAPIConnection with duckdb and sqlite when both support sqlalchemy?

Reply:
I realized that duckdb+sqlalchemy was generating a lot of issues, so we'll keep supporting it for retro compatibility but we'll now recommend users to use native connections with duckdb

Reply:
it isn't really that DBAPIConnection optimizes connections, it depends on the driver.